### DISCUSSION TITLE
Adding extra build information to executables.

### DISCUSSION THREAD

**Discussion by TheNeikos** [ipetkov/crane#35] [Q&A] [answered]
I am currently making an `egui`/`eframe` app, and they require the following inputs to run:

```nix
              xorg.libX11
              xorg.libXcursor
              xorg.libXrandr
              xorg.libXi
              libGL
              libGLU
```

I've tried to add them as buildInputs to all combinations of  `buildDepsOnly` and `buildPackage` but they never get found.
Additionally, the final executable(s) need a env variable (`LD_LIBRARY_PATH`) to run correctly.

Is there a way to leverage crane in this situation? I does find all executables in the workspace, so it could be great if one could specify a wrapper they need to run correctly?

Here is a link to a simple working devShell: 
https://github.com/TheNeikos/galleria/blob/763566df498b4ec194b8ad8c37af4899242c56a1/flake.nix#L87-L110
```nix
        devShell = pkgs.mkShell {
          inputsFrom = builtins.attrValues self.checks;


          buildInputs = with pkgs; [
            xorg.libX11
            xorg.libXcursor
            xorg.libXrandr
            xorg.libXi
            libGL
            libGLU
          ];


          # Extra inputs can be added here
          nativeBuildInputs = with pkgs; [
            rustToolchain
            cargo-tarpaulin
            cargo-edit
            pkgconfig
          ];


          shellHook = ''
            export LD_LIBRARY_PATH=/run/opengl-driver/lib/:${lib.makeLibraryPath ([pkgs.libGL pkgs.libGLU])}
          '';
        };
```

`nix run github:TheNeikos/galleria/763566df498b4ec194b8ad8c37af4899242c56a1#galleria`, fails with:

```
✦ ❯ nix run github:TheNeikos/galleria/763566df498b4ec194b8ad8c37af4899242c56a1#galleria
thread 'main' panicked at 'Failed to initialize any backend! Wayland status: NoCompositorListening X11 status: LibraryOpenError(OpenError { kind: Library, detail: "opening library failed (libX11.so.6: cannot open shared object file: No such file or directory); opening library failed (libX11.so: cannot open shared object file: No such file or directory)" })', /sources/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/winit-0.26.1/src/platform_impl/linux/mod.rs:619:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by shanesveller** [comment, marked as answer]
I can't seem to run your app on the flake-provided devShell on my hardware (6900XT GPU, NixOS `21.11.20220520.cbd40c7`), so I can't get to a fully working solution, but I think something like this is a helpful incremental step:

```diff
diff --git a/flake.nix b/flake.nix
index a7e4883..415fe52 100644
--- a/flake.nix
+++ b/flake.nix
@@ -50,6 +50,18 @@
         galleria = craneLib.buildPackage {
           inherit cargoArtifacts src;
         };
+
+        wrappedGalleria = pkgs.symlinkJoin {
+          name = "galleria";
+          paths = [ galleria ];
+
+          buildInputs = [ pkgs.makeWrapper ];
+
+          postBuild = ''
+            wrapProgram $out/bin/galleria \
+              --prefix LD_LIBRARY_PATH=${lib.makeLibraryPath ([pkgs.libGL pkgs.libGLU])}
+          '';
+        };
       in
       {
         checks = {
@@ -79,6 +91,7 @@
         };
 
         packages.galleria = galleria;
+        packages.wrapped = wrappedGalleria;
 
         apps.galleria = flake-utils.lib.mkApp {
           drv = galleria;
```

https://nixos.org/manual/nixpkgs/stable/#fun-makeWrapper
https://nixos.org/manual/nixpkgs/stable/#fun-wrapProgram
https://github.com/NixOS/nixpkgs/blob/3e7f627268fe561b8383a57d09deec1219c3db89/pkgs/build-support/trivial-builders.nix#L430
https://github.com/NixOS/nixpkgs/blob/3e7f627268fe561b8383a57d09deec1219c3db89/pkgs/build-support/setup-hooks/make-wrapper.sh#L13-L33

**Reply by TheNeikos** [reply]
Thank you very much! That showed me enough so that I could figure it out myself :D

It should now work on your machine as well: `nix run github:TheNeikos/galleria/7bdbea7#galleria`

Thanks!
